### ISSUE TITLE
Order of binding a field and value change listener differs the output.

### ISSUE THREAD

**Issue by abdullahtellioglu** [vaadin/flow#15295]
### Description of the bug

When I use binder and add a value change listener to a field, order matters which method is used first.  

If binding used before valueChangeListener `getBinder().getBean().getFoo()` equals same to `event.getValue();` at the callback event.

If addValueChangeListener is invoked before binding, `getBinder().getBean().getFoo()` returns the previous value at the callback event.

### Expected behavior

ValueChangeListener and binder's order should not effect this. If valueChangeListener adds to a field, this listener should be executed after binding operation is done.

```
addValueChangeListener();
binder.bind();
```
should work exactly like
```
binder.bind();
addValueChangeListener();
```

### Minimal reproducible example

```
@PageTitle("Main")
@Route(value = "")
public class MainView extends VerticalLayout {
    private RadioButtonGroup<String> radioButtonGroup;
    private final Span actualSelectedSpan = new Span();
    private final Span expectedSpan = new Span();
    private final Binder<FooPOJO> binder = new Binder<>(FooPOJO.class);

    public MainView() {
        binder.setBean(new FooPOJO());
        radioButtonGroup = new RadioButtonGroup<>("Sample Input Component", "Item1", "Item2", "Item3");
        add(radioButtonGroup);
        add(new Hr());
        add(actualSelectedSpan);
        add(expectedSpan);

        radioButtonGroup.addValueChangeListener(event -> {
            //prints the previous value all the time 
           actualSelectedSpan.setText("Actual = " + binder.getBean().getSelectedItem());
        });

        binder.forField(radioButtonGroup).bind("selectedItem");

        radioButtonGroup.addValueChangeListener(event -> {
             //prints the present value all the time 
            expectedSpan.setText("Expected = "+ binder.getBean().getSelectedItem());
        });                
    }
}

```

```
public class FooPOJO {
    private String selectedItem;
    public String getSelectedItem() {
        return selectedItem;
    }
    public void setSelectedItem(String selectedItem) {
        this.selectedItem = selectedItem;
    }
}
```

### Versions

- Vaadin / Flow version: 23.2.9
- Java version: 11
- OS version: 
- Browser version (if applicable):
- Application Server (if applicable):
- IDE (if applicable):

**Comment by Legioth** [comment]
To slightly simplify things, `binder.bind();` is just a shorthand for `addValueChangeListener();`.

If we want to make any guarantees about ordering, then we would have to add special API to all field components to have a "high priority" value change listener list exclusively for use by Binder and then a regular value change listener list for other use. Ordering in the other direction, so that binder only updates the bean after running other value change listeners can be done by making the operation asynchronous, but I'm not sure that makes sense either.

**Comment by mshabarov** [comment]
This new "high priority" API in binder shouldn't break the existing applications, so the order should be preserved, but be configurable.

**Comment by Legioth** [comment]
Good point. Backwards compatibility just makes this even more complicated. Sounds like all the complexity caused by this wouldn't be justifiable compared to the benefits. I could even claim that it's a good feature that you can choose at what point the listener is run by changing when it's added relative to setting up the binder.
